src/vhdl_backend.cc

```diff
@@ -0,0 +1,2130 @@
+			f << stringf(") begin\n");
+			ends = indent + "end\n" + ends;
+			indent += "  ";
+		}
+
+		if (sync->type == RTLIL::STp || sync->type == RTLIL::STn) {
+			for (size_t j = 0; j < proc->syncs.size(); j++) {
+				RTLIL::SyncRule *sync2 = proc->syncs[j];
+				if (sync2->type == RTLIL::ST0 || sync2->type == RTLIL::ST1) {
+					f << stringf("%s" "if (%s", indent.c_str(), sync2->type == RTLIL::ST1 ? "!" : "");
+					dump_sigspec(f, sync2->signal);
+					f << stringf(") begin\n");
+					ends = indent + "end\n" + ends;
+					indent += "  ";
+				}
+			}
+		}
+
+		for (auto it = sync->actions.begin(); it != sync->actions.end(); ++it) {
+			if (it->first.size() == 0)
+				continue;
+			f << stringf("%s  ", indent.c_str());
+			dump_sigspec(f, it->first);
+			f << stringf(" <= ");
+			dump_sigspec(f, it->second);
+			f << stringf(";\n");
+		}
+
+		f << stringf("%s", ends.c_str());
+	}
+}
+
+void dump_module(std::ostream &f, std::string indent, RTLIL::Module *module)
+{ // PORTING REQUIRED
+	reg_wires.clear();
+	reset_auto_counter(module);
+	active_module = module;
+	active_sigmap.set(module);
+	active_initdata.clear();
+	memory_array_types.clear();
+
+	for (auto wire : module->wires())
+		if (wire->attributes.count(ID::init)) {
+			SigSpec sig = active_sigmap(wire);
+			Const val = wire->attributes.at(ID::init);
+			for (int i = 0; i < GetSize(sig) && i < GetSize(val); i++)
+				if (val[i] == State::S0 || val[i] == State::S1)
+					active_initdata[sig[i]] = val[i];
+		}
+
+	// TODO: remove if unnecessary
+	if (!module->processes.empty())
+		log_warning("Module %s contains unmapped RTLIL processes. RTLIL processes\n"
+				"can't always be mapped directly to Verilog always blocks. Unintended\n"
+				"changes in simulation behavior are possible! Use \"proc\" to convert\n"
+				"processes to logic networks and registers.\n", log_id(module));
+
+	f << stringf("\n");
+	for (auto it = module->processes.begin(); it != module->processes.end(); ++it)
+		// This just updates internal data structures
+		dump_process(f, indent + "  ", it->second, true);
+
+	if (!noexpr)
+	{
+		std::set<std::pair<RTLIL::Wire*,int>> reg_bits;
+		for (auto cell : module->cells())
+		{
+			if (!reg_ct.count(cell->type) || !cell->hasPort(ID::Q))
+				continue;
+
+			RTLIL::SigSpec sig = cell->getPort(ID::Q);
+
+			if (sig.is_chunk()) {
+				RTLIL::SigChunk chunk = sig.as_chunk();
+				if (chunk.wire != NULL)
+					for (int i = 0; i < chunk.width; i++)
+						reg_bits.insert(std::pair<RTLIL::Wire*,int>(chunk.wire, chunk.offset+i));
+			}
+		}
+		for (auto wire : module->wires())
+		{
+			for (int i = 0; i < wire->width; i++)
+				if (reg_bits.count(std::pair<RTLIL::Wire*,int>(wire, i)) == 0)
+					goto this_wire_aint_reg;
+			if (wire->width)
+				reg_wires.insert(wire->name);
+		this_wire_aint_reg:;
+		}
+	}
+
+	// Entity declaration
+	// Find ports first before dumping them
+	std::map<int, Wire*> port_wires;
+	// TODO: validate assumption that port_id is port iff it is positive
```

Comment:
yosys codebase seems to use `if (wire->port_id != 0)` and `if (wire->port_input || wire->port_output)`, so one of those would probably be safer

Reply:
Going with `wire->port_input || wire->port_output` in commit 9c31e5b283ee311112b4d70f4903ef847e87dbc6.